### PR TITLE
Fix name-demangleling of provided types

### LINKED ISSUE #98: Using a single, optional, static parameter to a type provider causes failure
Originally opened on [CodePlex](https://visualfsharp.codeplex.com/workitem/71) by [latkin](Using a single, optional, static parameter to a type provider causes failure)
[originally reported to fsbugs] 

Repro solution attached 

2nhackv
If one creates a type provider with a single, optional, static parameter (i.e. single parameter with a default value specified), two scenarios are broken:
* Omitting the optional parameter 
* Including the optional parameter, when the value matches the specified default

E.g. TP "Test" has single string optional arg, with default value "Foo": 
Image 

Fix strategy, per Vlad: 
"Fix seems to be trivial - one line addition in name demangling routine (demangleProvidedTypeName in PrettyNaming.fs)"

[Repro](https://www.codeplex.com/Download/AttachmentDownload.ashx?ProjectName=visualfsharp&WorkItemId=71&FileAttachmentId=854201):

### PR DESCRIPTION
**Pull request by forki**
This fixes #98 

As reported in #98 the mangleling routine removes parameters which match the default value from the `typeLogicalName`. If all parameters match their default value we have a mangled name like `typeName,`  which is also reported:

image

I created a couple of unittests which pin the current behaviour. In order to do so I extracted the mangleling routine to make it testable.
I then added a testcase which tries to demangle `"MyNamespace.Test,"` and fails.

The last commit uses the strategy @vladima suggested in #98 and fixes the test case.

I think this will work fine, but I wonder if we should fix https://github.com/Microsoft/visualfsharp/blob/fsharp4/src/fsharp/PrettyNaming.fs#L466 instead and just don't generate logical type names with trailing `,`.

### CODE REVIEW

**Comment by forki** [comment]
I also wonder if this is the right place for the tests.

**Comment by forki** [comment]
I still think it would be better to not let the strange logical names happen in the first place. 
@dsyme what do you think?

File: src/fsharp/FSharp.Compiler.Unittests/FSharp.Compiler.Unittests.fsproj [hunk 1]
```diff
@@ -65,6 +65,7 @@
   </ItemGroup>
   <ItemGroup>
     <Compile Include="NUnitFrameworkShims.fs" Condition="'$(TargetFramework)' == 'sl3-wp'" />
+    <Compile Include="ManglelingNameOfProvidedTypes.fs" />
```
Comment (latkin) [inline src/fsharp/FSharp.Compiler.Unittests/FSharp.Compiler.Unittests.fsproj]: Proper spelling is "mangling"

File: src/fsharp/FSharp.Compiler.Unittests/ManglelingNameOfProvidedTypes.fs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+open System
+open System.Text
+
+open NUnit.Framework
+
+open Microsoft.FSharp.Compiler
+
+[<TestFixture>]
+type ManglelingNamesOfProvidedTypesWithSingleParameter() =
```
Comment (latkin) [inline src/fsharp/FSharp.Compiler.Unittests/ManglelingNameOfProvidedTypes.fs]: Throughout the file - include spaces between tupled elements/args

File: src/fsharp/PrettyNaming.fs [hunk 3]
```diff
@@ -445,16 +445,19 @@ module internal Microsoft.FSharp.Compiler.PrettyNaming
...
-                let argNamesAndValues = 
-                    pieces.[1..] |> Array.map (fun piece -> 
-                        match tryDemangleStaticStringArg piece with 
-                        | None -> raise (InvalidMangledStaticArg piece)
-                        | Some v -> v)
-                pieces.[0], argNamesAndValues
-            else 
-                typeLogicalName, [| |]
+            let pieces = splitAroundQuotation typeLogicalName ','
+            if pieces.Length = 2 && pieces.[1] = "" then
```
Comment (latkin) [inline src/fsharp/PrettyNaming.fs]: Could tighten this up a bit with pattern deconstruction and avoiding duplicate Some/None check.

```fsharp
match pieces with
| [| x; "" |] -> x, [| |]
| _ ->
    let argNamesAndValues = pieces.[1..] |> Array.choose tryDemangleStaticStringArg
    if argNamesAndValues.Length = (pieces.Length - 1) then
        pieces.[0], argNamesAndValues
    else
        typeLogicalName, [| |]
```

**Comment by latkin** [comment]
Looks nice. Just a couple cleanup suggestions.

**Comment by forki** [comment]
There is still the question if this is the right fix. 
Personally I'd prefer to not let the strange identifiers happen. I think I'll send a second pull request with alternative fix

**Comment by forki** [comment]
see https://github.com/Microsoft/visualfsharp/pull/203 for the second approach.

**Comment by forki** [comment]
closing this in favor of #203

**Pull request closed without merging**